fixed bug and implemented move food randomly after while (certain number of epochs) method

AlhaririAnas opened the pull request:
1. The input `new_pheromone_influence` should be a float type to avoid potential issues caused by incorrect error messages.


![grafik](https://github.com/heyitsalina/ant_search_algorithm/assets/147917604/e694f9a8-7405-4242-9c79-83cc5cb4f474)

2. A new method `move_randomly_after_while()` has been implemented within the Food class. This method moves the food source to a random location within the ant world after a certain number of epochs. Additionally, the amount of available food `Food.amount_of_food` is reset to its original value `Food.start_amount`. Furthermore, Food objects are removed from the list in the Simulation class once they are depleted. This enhances efficiency as non-existing food sources are no longer processed in the simulation.

![20240217-0340-32 9382060-ezgif com-video-to-gif-converter](https://github.com/heyitsalina/ant_search_algorithm/assets/147917604/7c8cdb7d-d73e-45ad-9e20-0c5f987da086)

Jonah-gr commented:
Interesting feature, but you must bear in mind that if the food objects are deleted from `simulation.food`, they will no longer appear in the created statistics and can therefore no longer be reloaded. That's why I implemented the statistic feature in such a way that the food objects are simply no longer displayed if their `amount_of_food == 0`, but are still present in `simulation.food`.

Nevertheless, we can of course implement this, but the user should always be able to switch the whole thing on and off. How did you intend to implement this?

AlhaririAnas commented:
> Interesting feature, but you must bear in mind that if the food objects are deleted from simulation.food, they will no longer appear in the created statistics and can therefore no longer be reloaded. That's why I implemented the statistic feature in such a way that the food objects are simply no longer displayed if their amount_of_food == 0, but are still present in simulation.food.
> 
> Nevertheless, we can of course implement this, but the user should always be able to switch the whole thing on and off. How did you intend to implement this?

Now, all Food objects are tracked in a new attribute, `Simulation.food_statistic`. The aim is to remove depleted Food objects (when Food.amount_of_food is 0) from `Simulation.food` list so they're no longer processed in the loop, and also to be able to reload the data of used-up food sources to.

I replaced the `checkpoint_epoch` parameter in the `Food.move_randomly_after_while()` method with a new attribute `Food.move_after_number_of_epochs` and added another attribute, `Food.move_randomly`: This decides if a Food object will be moved randomly. In the Food popup, you can set the needed number of epochs and use a switch to enable random movement individually.

Besides that, Food objects shouldn't be randomly placed on obstacles (regarding #132) to prevent ants from not reaching the food source. I'll soon work on ensuring Food objects do not overlap with obstacles.

What do you think about the new adjustes? @Jonah-gr

Jonah-gr commented:
I understand correctly that your problem is that you don't want to simply leave all food objects in 'Simulation.Food' for reasons of efficiency?
Because otherwise there's really no reason to delete the food objects from there, is there?

I didn't think your first implementation was bad at all, I was just wondering why food objects were being deleted. Instead of deleting them, wouldn't it be better to simply not place the food objects randomly if their `amount_of_food` is 0?
This would then include them again in a study and the last coordinates of the not yet "empty" food objects.

I think it makes a lot of sense to be able to switch these things on and off via the gui and set their parameters, I can help with that too.
In my opinion, this would have hardly any influence on the performance and would still allow the feature.

I hope I have understood you and you understand me correctly, otherwise feel free to ask further questions :)

AlhaririAnas commented:
> I understand correctly that your problem is that you don't want to simply leave all food objects in 'Simulation.Food' for reasons of efficiency?

> Instead of deleting them, wouldn't it be better to simply not place the food objects randomly if their amount_of_food is 0?

Exactly, for efficiency reasons, it makes sense to remove depleted Food objects from `Simulation.food`.

Our implementation with nested loops is quite complex:
```python
    def next_epoch(self):
        self.epoch += 1
        for colony in self.colonies:
            for ant in colony.ants:
                for food in self.food:
                    self.food_statistic.append(food)
                    if ant.try_carry_food(food):
                        ant.carry_food(food)
                        break
                if ant.try_drop_food(colony):
                    ant.drop_food(colony)
....
```

Let's say we have 3 colonies with 100 ants each and 10 Food objects, out of which 5 are already consumed. Each call to `next_epoch()` leads to 3 * 100 * 5 = 1500 unnecessary iterations, which is inefficient.

The runtime might not be severely affected by this, but it's still unnecessary.

Would it be problematic if we transfer the Food objects, including the depleted ones, into a separate attribute for study (statistc)?

Jonah-gr commented:
I see. I just think it's a bit confusing to create another attribute with food objects.
But as far as I can see, this really mainly deals with the iterations in `Simulation.next_epoch()`, right?
Couldn't we implement something simpler like this here:

```python
  def next_epoch(self):
      self.epoch += 1
      active_food_objects = [food for food in sim.food if food.amount_of_food != 0]
      for colony in self.colonies:
          for ant in colony.ants:
              for food in active_food_objects:
                  self.food_statistic.append(food)
                  if ant.try_carry_food(food):
                      ant.carry_food(food)
                      break
              if ant.try_drop_food(colony):
                  ant.drop_food(colony)
```

Would that be a solution/compromise?

AlhaririAnas commented:
> I see. I just think it's a bit confusing to create another attribute with food objects.
> But as far as I can see, this really mainly deals with the iterations in `Simulation.next_epoch()`, right?
> Couldn't we implement something simpler like this here:
> 
> ```python
>   def next_epoch(self):
>       self.epoch += 1
>       active_food_objects = [food for food in sim.food if food.amount_of_food != 0]
>       for colony in self.colonies:
>           for ant in colony.ants:
>               for food in active_food_objects:
>                   self.food_statistic.append(food)
>                   if ant.try_carry_food(food):
>                       ant.carry_food(food)
>                       break
>               if ant.try_drop_food(colony):
>                   ant.drop_food(colony)
> ```
> 
> Would that be a solution/compromise?


Yes, I agree with you, this is even better than creating a separate attribute. I will adopt your suggested solution. Thank you :)

AlhaririAnas commented:
ready to be merged

Jonah-gr commented:
Should I add the new features to the food dialog before merging? @AlhaririAnas

AlhaririAnas commented:
> Should I add the new features to the food dialog before merging? @AlhaririAnas

Oh! yes please, it would be better. I forgot about it.

Jonah-gr commented:
I hope you like that:

![image](https://github.com/heyitsalina/ant_search_algorithm/assets/119052351/91305e52-ba4a-421f-b528-09e7e0570ca5)

AlhaririAnas commented:
Very good! That's exactly how I imagined it 😄 

So i think it's ready to merge now
